contrib/redis/go-redis.v9/redis.go

```diff
@@ -0,0 +1,185 @@
+			additionalTags = []ddtrace.StartSpanOption{
+				tracer.Tag("out.db", strconv.Itoa(opt.DB)),
+			}
+		} else {
+			host, port, err := net.SplitHostPort(opt.Addr)
+			if err != nil {
+				host = opt.Addr
+				port = "6379"
+			}
+			additionalTags = []ddtrace.StartSpanOption{
+				tracer.Tag(ext.TargetHost, host),
+				tracer.Tag(ext.TargetPort, port),
+				tracer.Tag("out.db", strconv.Itoa(opt.DB)),
+			}
+		}
+	} else if clientOptions, ok := client.(clusterOptions); ok {
+		addrs := []string{}
+		for _, addr := range clientOptions.Options().Addrs {
+			addrs = append(addrs, addr)
+		}
+		additionalTags = []ddtrace.StartSpanOption{
+			tracer.Tag("addrs", strings.Join(addrs, ", ")),
+		}
+	}
+	return additionalTags
+}
+
+func (ddh *datadogHook) DialHook(hook redis.DialHook) redis.DialHook {
+	return func(ctx context.Context, network, addr string) (net.Conn, error) {
+		conn, err := hook(ctx, network, addr)
+		return conn, err
+	}
+}
+
+func (ddh *datadogHook) ProcessHook(hook redis.ProcessHook) redis.ProcessHook {
+	return func(ctx context.Context, cmd redis.Cmder) error {
+		raw := cmd.String()
+		length := strings.Count(raw, " ")
+		p := ddh.params
+		startOpts := make([]ddtrace.StartSpanOption, 0, 7+1+len(ddh.additionalTags)+1) // 7 options below + redis.raw_command + ddh.additionalTags + analyticsRate
+		startOpts = append(startOpts,
+			tracer.SpanType(ext.SpanTypeRedis),
+			tracer.ServiceName(p.config.serviceName),
+			tracer.ResourceName(raw[:strings.IndexByte(raw, ' ')]),
+			tracer.Tag("redis.args_length", strconv.Itoa(length)),
+			tracer.Tag(ext.Component, "redis/go-redis.v9"),
+			tracer.Tag(ext.SpanKind, ext.SpanKindClient),
+			tracer.Tag(ext.DBSystem, ext.DBSystemRedis),
+		)
+		if !p.config.skipRaw {
+			startOpts = append(startOpts, tracer.Tag("redis.raw_command", raw))
+		}
+		startOpts = append(startOpts, ddh.additionalTags...)
+		if !math.IsNaN(p.config.analyticsRate) {
+			startOpts = append(startOpts, tracer.Tag(ext.EventSampleRate, p.config.analyticsRate))
+		}
+		span, ctx := tracer.StartSpanFromContext(ctx, "redis.command", startOpts...)
+
+		err := hook(ctx, cmd)
+
+		var finishOpts []ddtrace.FinishOption
+		if err != nil && err != redis.Nil {
+			finishOpts = append(finishOpts, tracer.WithError(err))
+		}
+		span.Finish(finishOpts...)
+		return err
+	}
+}
+
+func (ddh *datadogHook) ProcessPipelineHook(hook redis.ProcessPipelineHook) redis.ProcessPipelineHook {
+	return func(ctx context.Context, cmds []redis.Cmder) error {
+		raw := commandsToString(cmds)
+		length := strings.Count(raw, " ")
+		p := ddh.params
+		startOpts := make([]ddtrace.StartSpanOption, 0, 8+1+len(ddh.additionalTags)+1) // 8 options below + redis.raw_command + ddh.additionalTags + analyticsRate
+		startOpts = append(startOpts,
+			tracer.SpanType(ext.SpanTypeRedis),
+			tracer.ServiceName(p.config.serviceName),
+			tracer.ResourceName(raw[:strings.IndexByte(raw, ' ')]),
+			tracer.Tag("redis.args_length", strconv.Itoa(length)),
```

Comment:
This tag doesn't make much sense for a Pipeline. Unless I'm mistaken, this seems to calculate all the spaces in the serialized forms of all of the commands in the pipeline. There could be multiple commands, each with a different number of arguments.
Furthermore, supporting this requires serializing all the commands into one big string. This causes unnecessary allocations and if the commands have particularly large payloads will cause a lot of thrashing in the GC.
The `SkipRaw` option doesn't actually prevent all this serialization from taking place.

Reply:
fixed in [71e6f68](https://github.com/DataDog/dd-trace-go/pull/1730/commits/71e6f680f1d3c304bed1249d3028e4029fb8130e)

Reply:
The above fix did not fully reflect the intent of your comment.
The following modification prevents unnecessary serialization when SkipRaw is specified.
a0c68d6